File: src/coding/system-prompt.ts
import type { ToolDefinition } from "../agent/index.js";
import type { ProjectInstruction } from "./project-context.js";

export interface SystemPromptOptions {
  contextFiles?: readonly ProjectInstruction[];
  cwd: string;
  tools: readonly ToolDefinition[];
}

function collectUnique(values: Iterable<string>): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const value of values) {
    const normalized = value.trim();
    if (normalized.length === 0 || seen.has(normalized)) continue;
    seen.add(normalized);
    result.push(normalized);
  }
  return result;
}

function escapeAttribute(value: string): string {
  return value
    .replaceAll("&", "&amp;")
    .replaceAll('"', "&quot;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;");
}

function formatProjectContext(files: readonly ProjectInstruction[]): string | undefined {
  if (files.length === 0) return undefined;
  const lines = ["<project_context>", "Project-specific instructions and guidelines:", ""];
  for (const file of files) {
    lines.push(`<project_instructions path="${escapeAttribute(file.path)}">`);
    lines.push(file.content);
    lines.push("</project_instructions>", "");
  }
  lines.push("</project_context>");
  return lines.join("\n");
}

export function buildSystemPrompt(options: SystemPromptOptions): string {
  const snippets = collectUnique(
    options.tools.flatMap((tool) => (tool.promptSnippet === undefined ? [] : [tool.promptSnippet])),
  );
  const guidelines = collectUnique(options.tools.flatMap((tool) => tool.promptGuidelines ?? []));

  const sections = [
    "You are Forge, a coding agent. Work directly on the user's task and report the result concisely.",
    `The workspace is ${options.cwd}. File tools are restricted to this workspace.`,
    "Inspect relevant files before changing them. Prefer precise edits over replacing entire files. Run appropriate checks after changes.",
  ];

  if (snippets.length > 0) sections.push(snippets.join("\n"));
  if (guidelines.length > 0) {
    sections.push(`Tool guidelines:\n${guidelines.map((item) => `- ${item}`).join("\n")}`);
  }
  const projectContext = formatProjectContext(options.contextFiles ?? []);
  if (projectContext !== undefined) sections.push(projectContext);

  return sections.join("\n\n");
}

File: src/forge_coding/system_prompt.py
"""System prompt assembly for Forge coding sessions."""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass, field
from datetime import date
from pathlib import Path
from xml.sax.saxutils import escape

from langchain_core.tools import BaseTool

from forge_coding.skills import Skill
from forge_coding.tools.definition import ToolDefinition
from forge_coding.tools.tool_set import ToolSet

SEQUENTIAL_TOOL_GUIDELINE = (
    "When a tool call depends on a previous tool result, wait for the next turn "
    "instead of batching both calls."
)


@dataclass(frozen=True, slots=True)
class ProjectContextFile:
    """A project instruction file included in the system prompt."""

    path: str
    content: str


@dataclass(frozen=True, slots=True)
class BuildSystemPromptOptions:
    """Options used to build Forge's system prompt."""

    cwd: Path
    tools: Sequence[BaseTool | ToolDefinition] | ToolSet = ()
    skills: Sequence[Skill] = ()
    custom_prompt: str | None = None
    append_system_prompt: str | None = None
    context_files: Sequence[ProjectContextFile] = ()
    current_date: date | None = None
    extra_guidelines: Sequence[str] = field(default_factory=tuple)


def build_system_prompt(options: BuildSystemPromptOptions) -> str:
    """Build a deterministic Pi-style system prompt for Forge."""
    current_date = options.current_date or date.today()
    cwd = _format_path(options.cwd)
    append_section = f"\n\n{options.append_system_prompt}" if options.append_system_prompt else ""

    if options.custom_prompt is not None:
        prompt = options.custom_prompt
        prompt += append_section
        prompt += format_project_context(options.context_files)
        if _has_tool(options.tools, "read"):
            prompt += format_skills_for_prompt(options.skills)
        prompt += f"\nCurrent date: {current_date.isoformat()}"
        prompt += f"\nCurrent working directory: {cwd}"
        return prompt

    prompt = (
        "You are an expert coding assistant operating inside Forge, a coding agent harness. "
        "You help users by reading files, executing commands, editing code, and writing new files."
        f"\n\nAvailable tools:\n{format_available_tools(options.tools)}"
        "\n\nIn addition to the tools above, you may have access to other custom tools "
        "depending on the project."
        f"\n\nGuidelines:\n{format_guidelines(options.tools, options.extra_guidelines)}"
    )

    prompt += append_section
    prompt += format_project_context(options.context_files)
    if _has_tool(options.tools, "read"):
        prompt += format_skills_for_prompt(options.skills)
    prompt += f"\nCurrent date: {current_date.isoformat()}"
    prompt += f"\nCurrent working directory: {cwd}"
    return prompt


def format_available_tools(tools: Sequence[BaseTool | ToolDefinition] | ToolSet) -> str:
    """Format visible tools using prompt snippets."""
    lines = [
        f"- {definition.name}: {definition.prompt_snippet}"
        for definition in _tool_definitions(tools)
        if definition.prompt_snippet
    ]
    return "\n".join(lines) if lines else "(none)"


def collect_prompt_guidelines(
    tools: Sequence[BaseTool | ToolDefinition] | ToolSet,
    extra_guidelines: Sequence[str] = (),
) -> list[str]:
    """Collect and de-duplicate system prompt guidelines."""
    definitions = _tool_definitions(tools)
    names = {definition.name for definition in definitions}
    guidelines: list[str] = []
    seen: set[str] = set()

    def add(value: str) -> None:
        normalized = value.strip()
        if not normalized or normalized in seen:
            return
        seen.add(normalized)
        guidelines.append(normalized)

    has_bash = "bash" in names
    has_exploration_tools = bool({"grep", "find", "ls"} & names)
    if has_bash and not has_exploration_tools:
        add("Use bash for file operations like ls, rg, find")
    elif has_bash and has_exploration_tools:
        add(
            "Prefer grep/find/ls tools over bash for file exploration (faster, respects .gitignore)"
        )

    for definition in definitions:
        for guideline in definition.prompt_guidelines:
            add(guideline)
    for guideline in extra_guidelines:
        add(guideline)

    add(SEQUENTIAL_TOOL_GUIDELINE)
    add("Be concise in your responses")
    add("Show file paths clearly when working with files")
    return guidelines


def format_guidelines(
    tools: Sequence[BaseTool | ToolDefinition] | ToolSet,
    extra_guidelines: Sequence[str] = (),
) -> str:
    """Format prompt guidelines as markdown bullets."""
    return "\n".join(
        f"- {guideline}" for guideline in collect_prompt_guidelines(tools, extra_guidelines)
    )


def format_project_context(context_files: Sequence[ProjectContextFile]) -> str:
    """Format project context files using Pi's XML-like wrapper."""
    if not context_files:
        return ""

    lines = [
        "\n\n<project_context>",
        "",
        "Project-specific instructions and guidelines:",
        "",
    ]
    for context_file in context_files:
        lines.append(f'<project_instructions path="{escape(context_file.path)}">')
        lines.append(context_file.content)
        lines.append("</project_instructions>")
        lines.append("")
    lines.append("</project_context>")
    return "\n".join(lines)


def format_skills_for_prompt(skills: Sequence[Skill]) -> str:
    """Format skills for inclusion in a system prompt using Pi's XML style."""
    if not skills:
        return ""

    lines = [
        "\n\nThe following skills provide specialized instructions for specific tasks.",
        "Read the full skill file when the task matches its description.",
        "When a skill file references a relative path, resolve it against the skill directory "
        "(parent of SKILL.md / dirname of the path) and use that absolute path in tool commands.",
        "",
        "<available_skills>",
    ]
    for skill in sorted(skills, key=lambda item: item.name):
        description = skill.description or "No description"
        lines.extend(
            [
                "  <skill>",
                f"    <name>{escape(skill.name)}</name>",
                f"    <description>{escape(description)}</description>",
                f"    <location>{escape(str(skill.path))}</location>",
                "  </skill>",
            ]
        )
    lines.append("</available_skills>")
    return "\n".join(lines)


def _has_tool(tools: Sequence[BaseTool | ToolDefinition] | ToolSet, name: str) -> bool:
    return any(definition.name == name for definition in _tool_definitions(tools))


def _tool_definitions(
    tools: Sequence[BaseTool | ToolDefinition] | ToolSet,
) -> tuple[ToolDefinition, ...]:
    """Normalize native tools and catalogs for prompt-only consumption."""

    if isinstance(tools, ToolSet):
        return tools.definitions
    return ToolSet.from_tools(tools).definitions


def _format_path(path: Path) -> str:
    return str(path).replace("\\", "/")
